File: app/src/main/java/com/neoruaa/xhsdn/LivePhotoCreator.java
package com.neoruaa.xhsdn;

import android.util.Log;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;

public class LivePhotoCreator {
    private static final String TAG = "LivePhotoCreator";

    /**
     * Creates a live photo by embedding video into image with XMP metadata
     * Based on the Motion Photo format specification
     * @param imageFile The image file to use as the primary content
     * @param videoFile The video file to embed
     * @param outputFile The output live photo file
     * @return True if successful, false otherwise
     */
    public static boolean createLivePhoto(File imageFile, File videoFile, File outputFile) {
        try {
            Log.d(TAG, "Creating live photo from image: " + imageFile.getAbsolutePath() + 
                   " (size: " + imageFile.length() + " bytes) and video: " + videoFile.getAbsolutePath() + 
                   " (size: " + videoFile.length() + " bytes) -> output: " + outputFile.getAbsolutePath());
            
            // Read the image and video files
            byte[] imageBytes = readFileToBytes(imageFile);
            byte[] videoBytes = readFileToBytes(videoFile);
            
            Log.d(TAG, "Read " + imageBytes.length + " bytes from image and " + videoBytes.length + " bytes from video");
            
            // Calculate the video start position after XMP insertion
            // First create the XMP to determine its size
            int videoOffset = imageBytes.length; // initial estimate
            String xmpData = generateXMPMetadata(videoBytes.length, videoOffset);
            byte[] xmpSegment = createXmpApp1Segment(xmpData);
            
            // Now we know the actual video offset after XMP insertion
            int actualVideoOffset = imageBytes.length + xmpSegment.length;
            
            // Recreate XMP with the correct offset
            String correctedXmpData = generateXMPMetadata(videoBytes.length, actualVideoOffset);
            byte[] correctedXmpSegment = createXmpApp1Segment(correctedXmpData);
            
            // Insert XMP right after SOI (0xFFD8) at position 2
            byte[] imageWithXmp = new byte[imageBytes.length + correctedXmpSegment.length];
            System.arraycopy(imageBytes, 0, imageWithXmp, 0, 2); // Copy SOI marker
            System.arraycopy(correctedXmpSegment, 0, imageWithXmp, 2, correctedXmpSegment.length); // Insert XMP
            System.arraycopy(imageBytes, 2, imageWithXmp, 2 + correctedXmpSegment.length, imageBytes.length - 2); // Copy rest of image
            
            // Combine image with XMP + video at the end
            byte[] result = new byte[imageWithXmp.length + videoBytes.length];
            System.arraycopy(imageWithXmp, 0, result, 0, imageWithXmp.length);
            System.arraycopy(videoBytes, 0, result, imageWithXmp.length, videoBytes.length);
            
            // Write the live photo bytes to output file
            try (FileOutputStream fos = new FileOutputStream(outputFile)) {
                fos.write(result);
            }
            
            Log.d(TAG, "Successfully created live photo with total size: " + outputFile.length() + " bytes");
            
            // Verify that the created file is valid
            if (!isLivePhotoValid(outputFile)) {
                Log.e(TAG, "Created live photo is not valid - failed validation check");
                if (outputFile.exists()) {
                    outputFile.delete(); // Clean up invalid file
                }
                return false;
            }
            
            Log.d(TAG, "Live photo validation passed successfully");
            return true;
        } catch (Exception e) {
            Log.e(TAG, "Error creating live photo: " + e.getMessage());
            e.printStackTrace();
            // If the file was created but is invalid, delete it
            if (outputFile.exists()) {
                outputFile.delete();
            }
            return false;
        }
    }
    

    
    /**
     * Generates XMP metadata for live photo following Google's Motion Photo specification
     * @param videoSize The size of the embedded video in bytes
     * @param videoOffset The offset where the video starts in the file
     * @return XMP metadata string
     */
    private static String generateXMPMetadata(int videoSize, int videoOffset) {
        return String.format(
            "<x:xmpmeta xmlns:x=\"adobe:ns:meta/\" x:xmptk=\"Adobe XMP Core 5.1.0-jc003\">\n" +
            "<rdf:RDF xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n" +
            "<rdf:Description rdf:about=\"\" xmlns:GCamera=\"http://ns.google.com/photos/1.0/camera/\" xmlns:Container=\"http://ns.google.com/photos/1.0/container/\" xmlns:Item=\"http://ns.google.com/photos/1.0/container/item/\">\n" +
            "   <GCamera:MotionPhoto>1</GCamera:MotionPhoto>\n" +
            "   <GCamera:MotionPhotoVersion>1</GCamera:MotionPhotoVersion>\n" +
            "   <GCamera:MotionPhotoPresentationTimestampUs>0</GCamera:MotionPhotoPresentationTimestampUs>\n" +
            "   <GCamera:MicroVideo>1</GCamera:MicroVideo>\n" +
            "   <GCamera:MicroVideoVersion>1</GCamera:MicroVideoVersion>\n" +
            "   <GCamera:MicroVideoPresentationTimestampUs>0</GCamera:MicroVideoPresentationTimestampUs>\n" +
            "   <GCamera:MicroVideoOffset>%d</GCamera:MicroVideoOffset>\n" +
            "   <GCamera:MicroVideoDuration>1000000</GCamera:MicroVideoDuration>\n" +
            "   <Container:Directory>\n" +
            "      <rdf:Seq>\n" +
            "         <rdf:li rdf:parseType=\"Resource\">\n" +
            "            <Item:Mime>image/jpeg</Item:Mime>\n" +
            "            <Item:Semantic>Primary</Item:Semantic>\n" +
            "            <Item:Length>0</Item:Length>\n" +
            "         </rdf:li>\n" +
            "         <rdf:li rdf:parseType=\"Resource\">\n" +
            "            <Item:Mime>video/mp4</Item:Mime>\n" +
            "            <Item:Semantic>MotionPhoto</Item:Semantic>\n" +
            "            <Item:Length>%d</Item:Length>\n" +
            "         </rdf:li>\n" +
            "      </rdf:Seq>\n" +
            "   </Container:Directory>\n" +
            "</rdf:Description>\n" +
            "</rdf:RDF>\n" +
            "</x:xmpmeta>",
            videoOffset, videoSize
        );
    }
    
    /**
     * Creates an APP1 XMP segment with proper JPEG header
     * @param xmpData The XMP data as string
     * @return Byte array representing the APP1 XMP segment
     */
    private static byte[] createXmpApp1Segment(String xmpData) throws IOException {
        byte[] xmpPayload = xmpData.getBytes("UTF-8");
        byte[] xmpHeader = "http://ns.adobe.com/xap/1.0/\0".getBytes("UTF-8");
        
        // The APP1 segment format is: [0xFFE1] [Length: 2 bytes] [Header] [Payload]
        // The Length field should contain the size of everything after the length field
        // So Length = size of (header + payload)
        int lengthField = xmpHeader.length + xmpPayload.length;
        
        // Create the segment
        ByteArrayOutputStream segment = new ByteArrayOutputStream();
        
        // Write APP1 marker (0xFFE1)
        segment.write(0xFF);
        segment.write(0xE1);
        
        // Write length field (2 bytes, big-endian) - value is size of header + payload
        segment.write((lengthField >> 8) & 0xFF);
        segment.write(lengthField & 0xFF);
        
        // Write XMP header
        segment.write(xmpHeader);
        
        // Write XMP data
        segment.write(xmpPayload);
        
        return segment.toByteArray();
    }
    
    /**
     * Validates if the created live photo is valid
     * @param file The file to validate
     * @return true if valid, false otherwise
     */
    private static boolean isLivePhotoValid(File file) {
        try (FileInputStream fis = new FileInputStream(file)) {
            byte[] header = new byte[10];
            int bytesRead = fis.read(header);
            if (bytesRead < 2) {
                return false;
            }
            
            // Check for JPEG SOI marker (0xFFD8)
            if (header[0] != (byte) 0xFF || header[1] != (byte) 0xD8) {
                Log.d(TAG, "File does not have valid JPEG SOI marker");
                return false;
            }
            
            // Check for the presence of XMP metadata by reading first part of file
            byte[] buffer = new byte[16384]; // Read first 16KB to look for XMP
            fis.close(); // Close the first stream
            
            // Read data to check for XMP metadata
            try (FileInputStream fis2 = new FileInputStream(file)) {
                int totalRead = 0;
                while (totalRead < buffer.length && (bytesRead = fis2.read(buffer, totalRead, buffer.length - totalRead)) != -1) {
                    totalRead += bytesRead;
                }
                
                String content = new String(buffer, 0, totalRead, "UTF-8");
                
                // Look for XMP signatures
                boolean hasXmpMeta = content.contains("xmpmeta");
                boolean hasMotionPhoto = content.contains("MotionPhoto");
                boolean hasMicroVideo = content.contains("MicroVideo");
                
                Log.d(TAG, "XMP validation - xmpmeta: " + hasXmpMeta + ", MotionPhoto: " + hasMotionPhoto + ", MicroVideo: " + hasMicroVideo);
                
                if (!hasXmpMeta || !hasMotionPhoto) {
                    Log.d(TAG, "File does not contain valid XMP Motion Photo metadata");
                    return false;
                }
            }
            
            // Try to decode the image to make sure it's valid
            android.graphics.BitmapFactory.Options options = new android.graphics.BitmapFactory.Options();
            options.inJustDecodeBounds = true;
            android.graphics.BitmapFactory.decodeFile(file.getAbsolutePath(), options);
            
            if (options.outWidth <= 0 || options.outHeight <= 0) {
                Log.d(TAG, "Image has invalid dimensions: " + options.outWidth + "x" + options.outHeight);
                return false;
            }
            
            return true;
            
        } catch (Exception e) {
            Log.e(TAG, "Error validating live photo: " + e.getMessage());
            return false;
        }
    }
    
    /**
     * Reads a file into a byte array
     * @param file The file to read
     * @return Byte array containing the file contents
     * @throws IOException
     */
    private static byte[] readFileToBytes(File file) throws IOException {
        try (FileInputStream fis = new FileInputStream(file);
             ByteArrayOutputStream buffer = new ByteArrayOutputStream()) {
            
            int nRead;
            byte[] data = new byte[16384]; // 16KB buffer
            
            while ((nRead = fis.read(data, 0, data.length)) != -1) {
                buffer.write(data, 0, nRead);
            }
            
            return buffer.toByteArray();
        }
    }
}

File: app/src/main/java/com/neoruaa/xhsdn/LivePhotoCreator.kt
package com.neoruaa.xhsdn

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.media.ExifInterface
import android.media.MediaScannerConnection
import android.util.Log
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.util.Locale

/** Creates Motion Photo/Live Photo files by prepending XMP metadata to a JPEG. */
object LivePhotoCreator {
    private const val TAG = "LivePhotoCreator"

    /**
     * Creates a live photo by embedding video into image with XMP metadata.
     */
    @JvmStatic
    fun createLivePhoto(
        imageFile: File,
        videoFile: File,
        outputFile: File,
        context: Context?,
    ): Boolean {
        val jpegFile = File(
            imageFile.parentFile,
            imageFile.name.replace(Regex("\\.[^.]+$"), "") + "_converted.jpg",
        )
        try {
            Log.d(
                TAG,
                "Creating live photo from image: ${imageFile.absolutePath} " +
                    "(size: ${imageFile.length()} bytes) and video: ${videoFile.absolutePath} " +
                    "(size: ${videoFile.length()} bytes) -> output: ${outputFile.absolutePath}",
            )

            Log.d(TAG, "Converting image to JPEG: ${jpegFile.absolutePath}")
            if (!convertToJpeg(imageFile, jpegFile)) {
                Log.e(TAG, "Failed to convert image to JPEG")
                return false
            }
            Log.d(TAG, "Successfully converted to JPEG: ${jpegFile.absolutePath} (size: ${jpegFile.length()} bytes)")

            val videoSize = videoFile.length()
            val xmpData = generateXmpMetadata(videoSize.toInt(), videoSize.toInt()).toByteArray(Charsets.UTF_8)
            val xmpSegment = createXmpApp1Segment(xmpData)
            val result = createLivePhotoStreaming(jpegFile, videoFile, outputFile, xmpSegment)

            if (result && context != null) {
                triggerMediaStoreScan(context, outputFile)
            }
            return result
        } catch (e: Exception) {
            Log.e(TAG, "Error creating live photo: ${e.message}", e)
            if (outputFile.exists() && !outputFile.delete()) {
                Log.w(TAG, "Failed to delete invalid output: ${outputFile.absolutePath}")
            }
            return false
        } finally {
            if (jpegFile.exists() && !jpegFile.delete()) {
                Log.w(TAG, "Failed to delete temporary JPEG: ${jpegFile.absolutePath}")
            }
        }
    }

    private fun triggerMediaStoreScan(context: Context, file: File) {
        MediaScannerConnection.scanFile(
            context,
            arrayOf(file.absolutePath),
            arrayOf("image/jpeg"),
        ) { path, uri -> Log.d(TAG, "Scanned: $path -> $uri") }
    }

    /** Converts any Android-decodable image to an orientation-normalized JPEG. */
    private fun convertToJpeg(inputFile: File, jpegFile: File): Boolean {
        var bitmap: Bitmap? = null
        var normalizedBitmap: Bitmap? = null
        return try {
            bitmap = BitmapFactory.decodeFile(inputFile.absolutePath)
            if (bitmap == null) {
                Log.e(TAG, "Failed to decode image: ${inputFile.absolutePath}")
                false
            } else {
                normalizedBitmap = normalizeBitmapOrientation(inputFile, bitmap)
                Log.d(
                    TAG,
                    "Decoded image: ${bitmap.width}x${bitmap.height}, " +
                        "normalized: ${normalizedBitmap.width}x${normalizedBitmap.height}",
                )
                FileOutputStream(jpegFile).use { output ->
                    normalizedBitmap.compress(Bitmap.CompressFormat.JPEG, 95, output)
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error converting to JPEG: ${e.message}")
            false
        } finally {
            if (normalizedBitmap != null && normalizedBitmap !== bitmap && !normalizedBitmap.isRecycled) {
                normalizedBitmap.recycle()
            }
            if (bitmap != null && !bitmap.isRecycled) {
                bitmap.recycle()
            }
        }
    }

    private fun normalizeBitmapOrientation(inputFile: File, bitmap: Bitmap): Bitmap {
        return try {
            val exif = ExifInterface(inputFile.absolutePath)
            val orientation = exif.getAttributeInt(
                ExifInterface.TAG_ORIENTATION,
                ExifInterface.ORIENTATION_NORMAL,
            )
            if (orientation == ExifInterface.ORIENTATION_UNDEFINED ||
                orientation == ExifInterface.ORIENTATION_NORMAL
            ) {
                return bitmap
            }

            val matrix = Matrix()
            when (orientation) {
                ExifInterface.ORIENTATION_FLIP_HORIZONTAL -> matrix.setScale(-1f, 1f)
                ExifInterface.ORIENTATION_ROTATE_180 -> matrix.setRotate(180f)
                ExifInterface.ORIENTATION_FLIP_VERTICAL -> matrix.setScale(1f, -1f)
                ExifInterface.ORIENTATION_TRANSPOSE -> {
                    matrix.setRotate(90f)
                    matrix.postScale(-1f, 1f)
                }
                ExifInterface.ORIENTATION_ROTATE_90 -> matrix.setRotate(90f)
                ExifInterface.ORIENTATION_TRANSVERSE -> {
                    matrix.setRotate(270f)
                    matrix.postScale(-1f, 1f)
                }
                ExifInterface.ORIENTATION_ROTATE_270 -> matrix.setRotate(270f)
                else -> return bitmap
            }

            val normalized = Bitmap.createBitmap(
                bitmap,
                0,
                0,
                bitmap.width,
                bitmap.height,
                matrix,
                true,
            )
            Log.d(
                TAG,
                "Applied EXIF orientation $orientation " +
                    "(rotation=${ImageOrientationUtils.rotationDegrees(orientation)}, " +
                    "swapsDimensions=${ImageOrientationUtils.swapsWidthAndHeight(orientation)})",
            )
            normalized
        } catch (e: IOException) {
            Log.w(TAG, "Failed to read EXIF orientation, using original bitmap: ${e.message}")
            bitmap
        }
    }

    @Suppress("unused")
    private fun isWebPFormat(file: File): Boolean {
        return try {
            FileInputStream(file).use { input ->
                val header = ByteArray(12)
                if (input.read(header) < header.size) {
                    false
                } else {
                    header[0] == 'R'.code.toByte() &&
                        header[1] == 'I'.code.toByte() &&
                        header[2] == 'F'.code.toByte() &&
                        header[3] == 'F'.code.toByte() &&
                        header[8] == 'W'.code.toByte() &&
                        header[9] == 'E'.code.toByte() &&
                        header[10] == 'B'.code.toByte() &&
                        header[11] == 'P'.code.toByte()
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error checking WebP format: ${e.message}")
            false
        }
    }

    @Suppress("unused")
    private fun convertWebPToJpeg(webpFile: File, jpegFile: File): Boolean {
        var bitmap: Bitmap? = null
        return try {
            bitmap = BitmapFactory.decodeFile(webpFile.absolutePath)
            if (bitmap == null) {
                Log.e(TAG, "Failed to decode WebP image")
                false
            } else {
                FileOutputStream(jpegFile).use { output ->
                    bitmap.compress(Bitmap.CompressFormat.JPEG, 95, output)
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error converting WebP to JPEG: ${e.message}")
            false
        } finally {
            if (bitmap != null && !bitmap.isRecycled) bitmap.recycle()
        }
    }

    private fun generateXmpMetadata(videoSize: Int, videoLengthForOffset: Int): String =
        String.format(
            Locale.ROOT,
            "<x:xmpmeta xmlns:x=\"adobe:ns:meta/\" x:xmptk=\"Adobe XMP Core 5.1.0-jc003\">" +
                "<rdf:RDF xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">" +
                "<rdf:Description rdf:about=\"\"" +
                "    xmlns:GCamera=\"http://ns.google.com/photos/1.0/camera/\"" +
                "    xmlns:OpCamera=\"http://ns.oplus.com/photos/1.0/camera/\"" +
                "    xmlns:MiCamera=\"http://ns.xiaomi.com/photos/1.0/camera/\"" +
                "    xmlns:Container=\"http://ns.google.com/photos/1.0/container/\"" +
                "    xmlns:Item=\"http://ns.google.com/photos/1.0/container/item/\"" +
                "  GCamera:MotionPhoto=\"1\"" +
                "  GCamera:MotionPhotoVersion=\"1\"" +
                "  GCamera:MotionPhotoPresentationTimestampUs=\"0\"" +
                "  OpCamera:MotionPhotoPrimaryPresentationTimestampUs=\"0\"" +
                "  OpCamera:MotionPhotoOwner=\"xhs\"" +
                "  OpCamera:OLivePhotoVersion=\"2\"" +
                "  OpCamera:VideoLength=\"%d\"" +
                "  GCamera:MicroVideoVersion=\"1\"" +
                "  GCamera:MicroVideo=\"1\"" +
                "  GCamera:MicroVideoOffset=\"%d\"" +
                "  GCamera:MicroVideoPresentationTimestampUs=\"0\"" +
                "  MiCamera:XMPMeta=\"&lt;?xml version='1.0' encoding='UTF-8' standalone='yes' ?&gt;\">" +
                "  <Container:Directory>" +
                "    <rdf:Seq>" +
                "      <rdf:li rdf:parseType=\"Resource\">" +
                "        <Container:Item" +
                "          Item:Mime=\"image/jpeg\"" +
                "          Item:Semantic=\"Primary\"" +
                "          Item:Length=\"0\"" +
                "          Item:Padding=\"0\"/>" +
                "      </rdf:li>" +
                "      <rdf:li rdf:parseType=\"Resource\">" +
                "        <Container:Item" +
                "          Item:Mime=\"video/mp4\"" +
                "          Item:Semantic=\"MotionPhoto\"" +
                "          Item:Length=\"%d\"/>" +
                "      </rdf:li>" +
                "    </rdf:Seq>" +
                "  </Container:Directory>" +
                "</rdf:Description>" +
                "</rdf:RDF>" +
                "</x:xmpmeta>",
            videoSize,
            videoSize,
            videoSize,
        )

    private fun createXmpApp1Segment(xmpData: ByteArray): ByteArray {
        val xmpHeader = "http://ns.adobe.com/xap/1.0/\u0000".toByteArray(Charsets.UTF_8)
        val segmentLength = xmpHeader.size + xmpData.size + 2
        return ByteArrayOutputStream().apply {
            write(0xFF)
            write(0xE1)
            write((segmentLength shr 8) and 0xFF)
            write(segmentLength and 0xFF)
            write(xmpHeader)
            write(xmpData)
        }.toByteArray()
    }

    private fun isLivePhotoValid(file: File): Boolean {
        return try {
            FileInputStream(file).use { input ->
                val header = ByteArray(10)
                val bytesRead = input.read(header)
                if (bytesRead < 2 || header[0] != 0xFF.toByte() || header[1] != 0xD8.toByte()) {
                    Log.d(TAG, "File does not have valid JPEG SOI marker")
                    return false
                }
            }

            val buffer = ByteArray(16 * 1024)
            var totalRead = 0
            FileInputStream(file).use { input ->
                while (totalRead < buffer.size) {
                    val count = input.read(buffer, totalRead, buffer.size - totalRead)
                    if (count == -1) break
                    totalRead += count
                }
            }
            val content = String(buffer, 0, totalRead, Charsets.UTF_8)
            val hasXmpMeta = content.contains("xmpmeta")
            val hasMotionPhoto = content.contains("MotionPhoto")
            val hasMicroVideo = content.contains("MicroVideo")
            Log.d(
                TAG,
                "XMP validation - xmpmeta: $hasXmpMeta, MotionPhoto: $hasMotionPhoto, " +
                    "MicroVideo: $hasMicroVideo",
            )
            if (!hasXmpMeta || !hasMotionPhoto) return false

            val options = BitmapFactory.Options().apply { inJustDecodeBounds = true }
            BitmapFactory.decodeFile(file.absolutePath, options)
            if (options.outWidth <= 0 || options.outHeight <= 0) {
                Log.d(TAG, "Image has invalid dimensions: ${options.outWidth}x${options.outHeight}")
                return false
            }
            true
        } catch (e: Exception) {
            Log.e(TAG, "Error validating live photo: ${e.message}")
            false
        }
    }

    @Suppress("unused")
    private fun readFileToBytes(file: File): ByteArray =
        FileInputStream(file).use { input ->
            ByteArrayOutputStream().use { output ->
                val buffer = ByteArray(16 * 1024)
                while (true) {
                    val count = input.read(buffer)
                    if (count == -1) break
                    output.write(buffer, 0, count)
                }
                output.toByteArray()
            }
        }

    private fun createLivePhotoStreaming(
        imageFile: File,
        videoFile: File,
        outputFile: File,
        xmpSegment: ByteArray,
    ): Boolean {
        return try {
            FileInputStream(imageFile).use { imageStream ->
                FileInputStream(videoFile).use { videoStream ->
                    FileOutputStream(outputFile).use { outputStream ->
                        val header = ByteArray(2)
                        if (imageStream.read(header) != 2) {
                            Log.e(TAG, "Could not read image header")
                            return false
                        }
                        outputStream.write(header)
                        outputStream.write(xmpSegment)

                        val buffer = ByteArray(8192)
                        val totalImageBytes = (imageFile.length() - 2).coerceAtLeast(0).toInt()
                        var copiedImageBytes = 0
                        while (copiedImageBytes < totalImageBytes) {
                            val bytesToRead = minOf(buffer.size, totalImageBytes - copiedImageBytes)
                            val bytesRead = imageStream.read(buffer, 0, bytesToRead)
                            if (bytesRead == -1) break
                            outputStream.write(buffer, 0, bytesRead)
                            copiedImageBytes += bytesRead
                        }

                        var copiedVideoBytes = 0L
                        while (true) {
                            val bytesRead = videoStream.read(buffer)
                            if (bytesRead == -1) break
                            outputStream.write(buffer, 0, bytesRead)
                            copiedVideoBytes += bytesRead
                        }
                        outputStream.flush()
                        Log.d(
                            TAG,
                            "Successfully created live photo with streaming approach. " +
                                "Image bytes copied: $copiedImageBytes, Video bytes copied: $copiedVideoBytes, " +
                                "Total file size: ${outputFile.length()}",
                        )
                    }
                }
            }

            if (!isLivePhotoValid(outputFile)) {
                Log.e(TAG, "Created live photo is not valid - failed validation check")
                if (outputFile.exists()) outputFile.delete()
                false
            } else {
                Log.d(TAG, "Live photo validation passed successfully")
                true
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error in streaming live photo creation: ${e.message}", e)
            if (outputFile.exists()) outputFile.delete()
            false
        }
    }
}
